docs/design/features/bridging-constraints.md

```diff
@@ -0,0 +1,110 @@
+    {
+        // The T in Do is unconstrainted, but DoConstrained requires it to be a struct
+        // The run time check above ensures this.
+        // This will not compile as C# today because the constraints don't match.
+        // Today, it will also result in a verification exception at runtime when compiling the Do method.
+        DoConstrained<T>();
+    }
+}
+```
+
+The purpose of this document is to sketch out a possible implementation on the runtime side that would allow above code to work.
+
+This feature doesn't assume any kind of relaxation of constraint checks outside method bodies. E.g. allowing to derive from `class Base<T> where T : IFoo { }` with `class Derived<T> : Base<T> { }` to load for `Derived<IFoo>` is not in scope.
+
+## Runtime impact
+
+There are two ways how constraints are validated within the runtime right now:
+
+* Validating they match on uninstantiated types: done by comparing constraints on generic type parameters - if type parameter T is used to substitute U, the constraints need to be satisfied.
+* Validating they match on concrete instantiation: for example `class Foo<T> where T: struct { }`: `Foo<int>` satisfies constraints, `Foo<object>` doesn't.
+
+Validating on uninstantiated forms is done as part of the old verification logic. The runtime doesn't do this consistently - for example, it is done in calls in IL, but not for ldtoken in IL.
+
+Validating instantiated forms is done always so that a `MethodTable` for an invalid type cannot even be created.
+
+From the runtime perspective, it would be desirable to maintain the invariant that `MethodTable` for a generic type instantiation that doesn't meet its own constraints can never be loaded - an attempt to do so would continue throwing an exception. This has implications on how method bodies that may be violating constraints in certain branches can be compiled. Consider:
+
+```csharp
+//
+// This is pseudo-C# just so that I don't have to write IL.
+// Not expected to compile as C# in this shape, but imagine the straightforward lowering to IL.
+//
+interface IInterface { }
+
+struct Constrained<T> where T : IInterface { }
+
+static void Do<T>()
+{
+    if (typeof(T).IsAssignableFrom(typeof(IInterface))
+    {
+        Constrained<T> someLocal = default;
+        try
+        {
+            someLocal.ToString();
+        }
+        catch (GenericException<Constrained<T>>) { }
+    }
+}
+```
+
+The requirement of never having an invalid type instantiation within the runtime type system means that if we're compiling `Do<int>` (where we cannot allow `Constrained<int>` to exist), the code generator cannot ask questions about all locals (IL locals are scoped to the method, not to a basic block), exception handling regions, or any IL within the `IsAssignableFrom` check. While it may be reasonable to expect the `IsAssignableFrom` check gets optimized out, it would likely not happen for unoptimized code. Locals and EH regions pose additional programs.
```

Comment:
I am wondering whether we can limit the patterns for bridging the constraints to just method calls. If there is more than just a simple method call that needs to execute with the more stringent constraint, the code can be factored out into a separate (aggressively inlined) method.

Reply:
If we do that, we may not need the fine-grained runtime helpers for checking the constraints. We may limit the support to the following pattern:

```
call RuntimeHelpers.CheckConstraintsForMethodCallThatFollows
jne ....
push arg0
push arg1
...
push argN
call CodeWithMoreStringentConstraints
```

Reply:
Imagine in C# code, a branch may only be taken if a stronger constraint is met:
```csharp
interface IA {}
interface IB : IA {}
M<T>()
{
    if (T is IB) // stronger than IA
    {
        M2<T>();
    }
}

M2<T>() where T : IA {}
```

Reply:
The compiler can introduce a compiler-generated method with the stronger constraint in this case.